Add loadwallet and createwallet load_on_startup options

ryanofsky opened the pull request:
This maintains a persistent list of wallets stored in settings that will automatically be loaded on startup. Being able to load a wallet automatically on startup will be more useful in the GUI, but it's reasonable to expose this feature by RPC as well.

ryanofsky committed:
Add loadwallet and createwallet RPC load_on_startup options

This maintains a persistent list of wallets stored in settings that will
automatically be loaded on startup. Being able to load a wallet automatically
on startup will be more useful in the GUI when the option to create wallets is
added in #15006, but it's reasonable to expose this feature by RPC as well.

jonasschnelli commented:
Strong Concept ACK
"Loosing" the wallets especially in conjunction with pruning is a concept hard to understand.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2688,6 +2701,11 @@ static UniValue createwallet(const JSONRPCRequest& request)
     }
 
+    bool load_on_startup = false;
+    if (!request.params[4].isNull() && request.params[4].isBool()) {
```
Nit, remove ` && request.params[4].isBool()`?

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2626,10 +2628,20 @@ static UniValue loadwallet(const JSONRPCRequest& request)
     if (!wallet) throw JSONRPCError(RPC_WALLET_ERROR, error);
 
+    if (load_on_startup && !AddWalletSetting(wallet->chain(), location.GetName())) {
+        throw JSONRPCError(RPC_MISC_ERROR, "Wallet loaded, but load-on-startup list could not be written, so wallet "
```
An error is thrown but wallet remains loaded. Either:
 - make this a warning in the response;
 - move `AddWalletSetting()` before loading - which must be reverted if the actual load fails;
 - unload the wallet before throwing.

promag reviewed (commented):
> but it's reasonable to expose this feature by RPC as well

For GUI this makes perfect sense, for RPC I'm not sure if there's a use case that pays off.

Quickly reviewed just 9d5c780dcb8d573aee934d110a27194a637fed8d.

luke-jr commented:
>For GUI this makes perfect sense, for RPC I'm not sure if there's a use case that pays off.

bitcoind also autoloads wallets as configured. (And it only really makes sense in general to use bitcoin_rw.conf for this...)

achow101 commented:
Concept ACK. I think I'll rebase #15454 on top of this.

ryanofsky reviewed (commented):
re: https://github.com/bitcoin/bitcoin/pull/15937#pullrequestreview-254036448

> For GUI this makes perfect sense, for RPC I'm not sure if there's a use case that pays off.

Use case is not having to manually create, edit, restart and test with a bitcoin.conf file just to use multiple wallets. I also think the gui implementation will be simpler and better tested building on the rpc implementation (even if some code has to move).

achow101 commented:
Is it intended that the default wallet won't be loaded anymore when a wallet is explicitly loaded? That seems like it would cause problems for users who are using both the default wallet and other wallets.

ryanofsky commented:
> Is it intended that the default wallet won't be loaded anymore when a wallet is explicitly loaded? That seems like it would cause problems for users who are using both the default wallet and other wallets.

Good catch. No that's a bug. Should continue to load default wallet unless it was explicitly overridden with a -wallet or -nowallet argument or config value (as before). Will fix and add some tests.

> Also, why default to not loading on start? I think it makes sense that restarting would load all previously loaded wallets too.

I think that's be a natural followup change, and maybe a good thing to in conjunction with the GUI PR to make default RPC behavior align with GUI behavior.

Reason why it is not done here is because I would like this change to be backward compatible and not break existing workflows (or require updating lots of tests). This PR is just trying to add a simple feature to demonstrate loading and saving a setting, not be a big user facing change. But if anything here gets in the way of the larger feature, I'd be fine with closing this PR and just going for the big change.

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/15937#issuecomment-643475377

> > Is it intended that the default wallet won't be loaded anymore when a wallet is explicitly loaded? That seems like it would cause problems for users who are using both the default wallet and other wallets.
> 
> Good catch. No that's a bug. Should continue to load default wallet unless it was explicitly overridden with a -wallet or -nowallet argument or config value (as before). Will fix and add some tests.

Fixed. The existing test covered this case so I updated it and added some new tests.

Updated 36b3318a480eaa6fcc683b6f85ae60e809083faf -> 9b69729015eee6c70d0209396bd6d34024e2aed4 ([`pr/walset.7`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.7) -> [`pr/walset.8`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/walset.7..pr/walset.8)) with default wallet fix

achow101 commented:
ACK 6c6846cd58e33108887cea4e2a1127a56e8c7d67

achow101 commented:
Did a bit of testing, and it seems another issue is that `unloadwallet` will cause the default wallet to no longer be loaded as well. If multiple wallets are loaded with `-wallet=` and then one of those is unloaded, the setting will get written without the default wallet which causes it to also not be loaded.

ryanofsky commented:
> Did a bit of testing, and it seems another issue is that `unloadwallet` will cause the default wallet to no longer be loaded as well. If multiple wallets are loaded with `-wallet=` and then one of those is unloaded, the setting will get written without the default wallet which causes it to also not be loaded.

This was intentional. This idea is for this pr to be backwards compatible as long as you don't use use `load_on_startup=true`. Once you actively start using `load_on_startup=true`, it jettisons the old behavior where a default "" wallet you unloaded and didn't ask for can get created/loaded automatically on a future startup. If a default "" wallet was already loaded for you on startup and you didn't unload it, it will still keep being loaded (for now: after #15454 we can [drop this behavior](https://github.com/ryanofsky/bitcoin/blob/pr/walset.8/src/wallet/load.cpp#L132-L136) too). This gives some of the advantages from #15454 of being able to control what wallets are loaded on startup without having to edit a text file, but it keeps things backwards compatible until we roll out the more impactful #15454 changes.

If this isn't conservative enough, I can add a load_on_startup true/false/null argument to unloadwallet RPC, where the default is to not change the startup setting instead of removing unloaded wallets from the startup list. This might be useful if maybe there are use-cases for wanting to unload a wallet temporarily but set it to be reloaded or still keep loading it on startup.

> Additionally if a wallet is set to load on startup but is also specified in the command line, we will fail to start with a `Duplicate wallet file` error.

I think the error is actually informative and good, but I would like to improve the error message, because it's already ambiguous and confusing when its thrown from RPCs https://github.com/bitcoin/bitcoin/issues/19232#issuecomment-645114976. It could be improved here or in the other PR.

achow101 commented:
> Once you actively start using `load_on_startup=true`, it jettisons the old behavior where a default "" wallet you unloaded and didn't ask for can get created/loaded automatically on a future startup. If a default "" wallet was already loaded for you on startup and you didn't unload it,

I agree that's what should happen, but this issue was because I had somehow had a default wallet loaded on startup, unloaded a different wallet, and on restart, found that the default wallet was no longer being loaded. The expectation was that it would still be loaded because it had been loaded automatically earlier.

Although now I can't replicate this. I'm not sure what sequence of steps arrived at that.

achow101 commented on src/wallet/load.cpp:
```diff
@@ -117,3 +119,39 @@ void UnloadWallets()
+        // default SoftSetArg wallet is loaded, add the default wallet to the
+        // list so it still gets loaded next startup. This can be removed after
+        // https://github.com/bitcoin/bitcoin/pull/15454, when there is no more
+        // default SoftSetArg wallet.
```
I don't think we can remove this behavior in #15454 as we still are loading the default wallet if it exists. So we still need this behavior there.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/15937#discussion_r443866400

> I don't think we can remove this behavior in #15454 as we still are loading the default wallet if it exists. So we still need this behavior there.

Thanks, rethinking this and looking at #15454, there's no need to have a temporary workaround here at all if WalletInit::Construct initializes the wallet setting, so the PR now does that and drops this change.

ryanofsky reviewed (commented):
Rebased 9b69729015eee6c70d0209396bd6d34024e2aed4 -> cd2e0e090beb4c8f2ea61257eed76a7585643da0 ([`pr/walset.8`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.8) -> [`pr/walset.9`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.9), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/walset.8-rebase..pr/walset.9)) on top of #15935 pr/rwset.13, also including changes to make load_on_startup option more flexible and simplify handling of default wallet
Rebased cd2e0e090beb4c8f2ea61257eed76a7585643da0 -> 43f8faefc3e47b06ce20ffd0df85589cde7fa21d ([`pr/walset.9`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.9) -> [`pr/walset.10`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.10), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/walset.9-rebase..pr/walset.10)) due to conflict with #19331
Rebased 43f8faefc3e47b06ce20ffd0df85589cde7fa21d -> 1a73d115c4b9e9ede60822d6e9fa1421c81387f2 ([`pr/walset.10`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.10) -> [`pr/walset.11`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.11), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/walset.10-rebase..pr/walset.11)) on top of #15935 pr/rwset.14
Rebased 1a73d115c4b9e9ede60822d6e9fa1421c81387f2 -> 65bcf8e17a5a3c2324996d369415dde597a834e0 ([`pr/walset.11`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.11) -> [`pr/walset.12`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.12), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/walset.11-rebase..pr/walset.12)) on top of #15935 pr/rwset.15 due to #19493 travis failures https://travis-ci.org/github/bitcoin/bitcoin/builds/707135059 and conflict with #18923
Rebased 65bcf8e17a5a3c2324996d369415dde597a834e0 -> 4801c7f8f5dc8f6cf09e1afe8f928d267258d2cc ([`pr/walset.12`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.12) -> [`pr/walset.13`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.13), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/walset.12-rebase..pr/walset.13)) on top of #15935 pr/rwset.17, also improved documentation and added release notes
Rebased 4801c7f8f5dc8f6cf09e1afe8f928d267258d2cc -> 0b146a4d682edb30b52af151eacce76d526dc91d ([`pr/walset.13`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.13) -> [`pr/walset.14`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.14), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/walset.13-rebase..pr/walset.14)) after base #15935 merge
Updated 0b146a4d682edb30b52af151eacce76d526dc91d -> 3062303a682492f3c90e64eb008dbdfe54e0cb60 ([`pr/walset.14`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.14) -> [`pr/walset.15`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.15), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/walset.14..pr/walset.15)) fixing release note typo

jonatack commented:
Concept ACK.

I would definitely find this useful and use it.

meshcollider commented:
Concept ACK

achow101 commented on test/functional/wallet_startup.py:
```diff
@@ -0,0 +1,50 @@
+        self.nodes[0].loadwallet(filename='w4', load_on_startup=True)
+        assert_equal(set(self.nodes[0].listwallets()), set(('', 'w1', 'w2', 'w3', 'w4')))
+        self.stop_nodes()
+        self.start_node(0)
```
nit: this could be `self.restart_node(0)`

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/15937#discussion_r465921931

> nit: this could be `self.restart_node(0)`

Thanks, updated

achow101 reviewed (commented):
ACK 3062303a682492f3c90e64eb008dbdfe54e0cb60

promag reviewed (commented):
Tested ACK 3062303a682492f3c90e64eb008dbdfe54e0cb60. Played around and works as expected.

How would you see this option in the GUI? A checkbox somewhere?

Could also test with `load_on_startup` omitted.

achow101 commented:
> How would you see this option in the GUI? A checkbox somewhere?

For #15454, the GUI implementation is to just make this the default.

promag commented:
@achow101 need to review #15454 again. Do you mean load_on_startup=true for menus create and load? If so, I menu unload should set load_on_startup=false.

achow101 commented:
> Do you mean load_on_startup=true for menus create and load?

`load_on_startup=true` for loading, and `load_on_startup=false` on unloading.

ryanofsky commented:
Rebased 3062303a682492f3c90e64eb008dbdfe54e0cb60 -> 904bc013e36dc6ea9d013e664d5e9446aae94c7b ([`pr/walset.15`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.15) -> [`pr/walset.16`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.16), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/walset.15-rebase..pr/walset.16)) due to conflict with #18654
Updated 904bc013e36dc6ea9d013e664d5e9446aae94c7b -> 61709e87c6a92e12e8ecf4cdffc058ebb565ccdc ([`pr/walset.16`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.16) -> [`pr/walset.17`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.17), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/walset.16..pr/walset.17)) with test suggestion

achow101 commented:
re-ACK 61709e87c6a92e12e8ecf4cdffc058ebb565ccdc. Only changes since last was rebase and test suggestion.

meshcollider commented on test/functional/wallet_startup.py:
```diff
@@ -0,0 +1,48 @@
+        self.nodes[0].unloadwallet(wallet_name='', load_on_startup=False)
+        self.nodes[0].unloadwallet(wallet_name='w4', load_on_startup=False)
+        self.nodes[0].loadwallet(filename='w3', load_on_startup=True)
+        self.nodes[0].loadwallet(filename='', load_on_startup=False)
```
very minor suggestion: `load_on_startup` is already false for this wallet so it may be a more interesting test to check load without a value keeps that previous setting here.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/15937#discussion_r470259520

> very minor suggestion: `load_on_startup` is already false for this wallet so it may be a more interesting test to check load without a value keeps that previous setting here.

Good idea, updated

meshcollider reviewed (commented):
Code review ACK 61709e87c6a92e12e8ecf4cdffc058ebb565ccdc

ryanofsky reviewed (commented):
Updated 61709e87c6a92e12e8ecf4cdffc058ebb565ccdc -> 642ad31b418bbf8da06cb3641329b0810e18e55b ([`pr/walset.17`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.17) -> [`pr/walset.18`](https://github.com/ryanofsky/bitcoin/commits/pr/walset.18), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/walset.17..pr/walset.18)) with test suggestion

meshcollider reviewed (commented):
re-utACK 642ad31b418bbf8da06cb3641329b0810e18e55b

Only change is modification to test as suggested

achow101 commented:
re-ACK 642ad31b418bbf8da06cb3641329b0810e18e55b Only change is the test

promag commented:
> > Do you mean load_on_startup=true for menus create and load?
> 
> `load_on_startup=true` for loading, and `load_on_startup=false` on unloading.

Not sure if you interpret create as loading, but IMO created wallets should have load_on_startup=true too.

jonatack commented:
Posthumous strong concept ACK -- looking forward to using this feature.
